In [3]:
import pandas as pd
from pathlib import Path

RESULTS_DF_PATH = Path("../data/processed/model_metrics.csv")
metrics_df = pd.read_csv(RESULTS_DF_PATH, index_col=0)

In [4]:
# show metrics df
metrics_df.head()

,CV Mean R²,CV Std,Test R² score,MAE,MSE,MAPE
Dummy,-0.003,0.004,-0.001,60344.244,7.119421e+09,0.372
SVR,-0.063,0.021,-0.070,58086.537,7.610946e+09,0.319
Random Forest Regression,0.897,0.010,0.932,13821.311,4.847736e+08,0.078
XG Boost,0.901,0.012,0.934,13848.479,4.676566e+08,0.079


## Machine Learning Conclusions

I trained four regression models to predict house sale price, namely, a dummy regressor, Support Vector Machine (SVR), Random Forest and XG Boost. Some conclusions worth noting are:
1.	The cross-validation and test-set R² scores are generally consistent, suggesting that the models generalize well and are not severely overfitting. The best test scores were achieved by Random Forest and XG Boost, both achieving 0.93. Ensemble models had the advantage over the other models, performing well overall and handling nonlinear data effectively.
2.	It is also worth noting that, in many cases, SVR performed worse than the dummy regressor. The SVR model failed to capture the relationship between the predictors and sale price. Its negative R² indicates that a naive strategy of always predicting the mean sale price outperformed the model.
The mean absolute error (MAE) is slightly better, getting its predictions wrong by about 58k, instead of 60k. 
3.	The top models were Random Forest and XG Boost, which excelled during testing. Both show a strong ability to learn the pattern effectively, achieving high test R2 scores and achieving mean absolute errors of ≈$13.8k. Furthermore, the low mean absolute percentage error (MAPE) of 8% suggests that their predictions are off by only 8%. Random Forest and XGBoost produced nearly identical results, with XGBoost achieving a marginally higher R² score. The difference is small enough that both models can be considered top performers on this dataset.
Tree-based ensemble methods substantially outperformed both the baseline and SVR models, suggesting that the relationship between housing features and sale price is highly nonlinear.